apps/meteor/app/push/server/fcm.ts

```diff
@@ -0,0 +1,159 @@
+};
+
+type FCMMessage = {
+	notification?: FCMNotificationField;
+	data?: FCMDataField;
+	token?: string;
+	to?: string;
+	android?: {
+		collapseKey?: string;
+		priority?: 'HIGH' | 'NORMAL';
+		ttl?: string;
+		restrictedPackageName?: string;
+		data?: FCMDataField;
+		notification?: FCMNotificationField;
+		fcm_options?: {
+			analytics_label?: string;
+		};
+		direct_boot_ok?: boolean;
+	};
+	webpush?: {
+		headers?: FCMDataField;
+		data?: FCMDataField;
+		notification?: FCMNotificationField;
+		fcm_options?: {
+			link?: string;
+			analytics_label?: string;
+		};
+	};
+	fcm_options?: {
+		analytics_label?: string;
+	};
+};
+
+// https://firebase.google.com/docs/reference/fcm/rest/v1/ErrorCode
+type FCMError = {
+	error: {
+		code: number;
+		message: string;
+		status: string;
+	};
+};
+
+function fetchWithRetry(url: string, options: RequestInit, retries = 0): Promise<Response> {
+	const MAX_RETRIES = 5;
+	const ERRORS_TO_RETRY = [429, 500, 502, 503, 504];
+
+	try {
+		return fetch(url, options);
+	} catch (error) {
+		if (retries >= MAX_RETRIES) {
+			throw error;
+		}
+
+		const errorData = error as FCMError;
+
+		if (!errorData?.error?.code || !ERRORS_TO_RETRY.includes(errorData.error.code)) {
+			throw error;
+		}
+
+		return fetchWithRetry(url, options, retries + 1);
+	}
+}
+
+function getFCMMessagesFromPushData(userTokens: string[], notification: PendingPushNotification): { message: FCMMessage }[] {
+	// first we will get the `data` field from the notification
+	const data: FCMDataField = notification.payload ? { ejson: EJSON.stringify(notification.payload) } : {};
+
+	// Set image
+	if (notification.gcm?.image != null) {
+		data.image = notification.gcm?.image;
+	}
+
+	// Set extra details
+	if (notification.badge != null) {
+		data.msgcnt = notification.badge.toString();
+	}
+	if (notification.sound != null) {
+		data.soundname = notification.sound;
+	}
+	if (notification.notId != null) {
+		data.notId = notification.notId.toString();
+	}
+	if (notification.gcm?.style != null) {
+		data.style = notification.gcm?.style;
+	}
+
+	if (notification.contentAvailable != null) {
+		data['content-available'] = notification.contentAvailable.toString();
+	}
+
+	// then we will create the notification field
+	const notificationField: FCMNotificationField = {
+		title: notification.title,
+		body: notification.text,
+	};
+
+	// then we will create the message
+	const message: FCMMessage = {
+		notification: notificationField,
+		data,
+		android: {
+			priority: 'HIGH',
+		},
+	};
+
+	// then we will create the message for each token
+	return userTokens.map((token) => ({ message: { ...message, token } }));
+}
+
+export const sendFCM = function ({ userTokens, notification, _replaceToken, _removeToken, options }: NativeNotificationParameters): void {
+	// We don't use these parameters, but we need to keep them to keep the function signature
+	// TODO: Remove them when we remove the old sendGCM function
+	_replaceToken;
+	_removeToken;
```

Comment:
Why these references?

Reply:
Just to keep the interface compatible with the legacy `sendFcm` API
It should be removed whenever we delete that code (not sure when though, hope it lands on 7.0)

Reply:
M... May I ask you to explain why we don't need them? As I see, this PR does not contain a code to clear the mongo database from invalid tokens. Unlike sendGcm does.

Is it irrelevant now?

Reply:
Hey @Lao-Ax thanks for your comment, I was mistaken on that one, there is no need to remove the tokens if it failed, since it is impossible in the new API
But, it looks like the logic for replacing stale tokens is missing, I will open a quick PR to implement that, thanks again :)

Reply:
It looks like there is no way to tell if we should just refresh the token or invalidate it
but the spec says that if we get a 404 we should remove it from the database, good catch!